src/squidpy/im/_feature.py

```diff
@@ -1,20 +1,128 @@
 from __future__ import annotations
 
-from collections.abc import Mapping, Sequence
+import typing
+import warnings
+from collections.abc import Generator, Mapping, Sequence
 from types import MappingProxyType
-from typing import TYPE_CHECKING, Any
+from typing import TYPE_CHECKING, Any, Callable, TypeVar, Union
 
+import numpy as np
+import numpy.typing as npt
 import pandas as pd
+import spatialdata as sd
+import xarray as xr
 from anndata import AnnData
 from scanpy import logging as logg
+from skimage.measure import perimeter, regionprops
+from spatialdata import SpatialData
 
 from squidpy._constants._constants import ImageFeature
 from squidpy._docs import d, inject_docs
 from squidpy._utils import Signal, SigQueue, _get_n_cores, parallelize
 from squidpy.gr._utils import _save_data
+from squidpy.im import _measurements
 from squidpy.im._container import ImageContainer
 
-__all__ = ["calculate_image_features"]
+__all__ = ["calculate_image_features", "quantify_morphology"]
+
+IntegerNDArrayType = TypeVar("IntegerNDArrayType", bound=npt.NDArray[np.integer[Any]])
+FloatNDArrayType = TypeVar("FloatNDArrayType", bound=npt.NDArray[np.floating[Any]])
+RegionPropsCallableType = Callable[
+    [IntegerNDArrayType, FloatNDArrayType], Union[Union[int, float, list[Union[int, float]]]]
+]
+
+RegionPropsImageCallableType = Callable[[IntegerNDArrayType, FloatNDArrayType], dict[str, Union[int, float]]]
+
+
+def circularity(regionmask: IntegerNDArrayType) -> float:
+    """
+    Calculate the circularity of the region.
+
+    :param regionmask: Region properties object
+    :return: circularity of the region
+    """
+    perim = perimeter(regionmask)
+    if perim == 0:
+        return 0
+    area = np.sum(regionmask)
+    return float((4 * np.pi * area) / (perim**2))
+
+
+def _get_region_props(
+    label_element: xr.DataArray,
+    image_element: xr.DataArray,
+    props: list[str] | None = None,
+    extra_methods: list[RegionPropsCallableType] | None = None,
+) -> pd.DataFrame:
+    if not extra_methods:
+        extra_methods = []
+    if props is None:
+        # if we didn't get any properties, we'll do the bare minimum
+        props = ["label"]
+
+    np_rgb_image = image_element.values.transpose(1, 2, 0)  # (c, y, x) -> (y, x, c)
```

Comment:
like, stuff like this doesn't work with real data. Everything needs to be either lazy, or looped/parallelized